src/RevisionPopupWidget.js

```diff
@@ -146,11 +143,37 @@ RevisionPopupWidget.prototype.show = function ( data, $target, isCached ) {
 		$scorePercent = $( '<strong>' ).text( mw.msg( scoreMsgKey + '-percent', data.score ) ),
 		$attributionMsg = $( '<div>' )
 			.addClass( 'wwt-revisionPopupWidget-attribution' )
-			.html( Tools.i18nHtml( scoreMsgKey, $scorePercent ) ),
-		$html = $( '<div>' )
-			.append( addedMsg, getCommentHtml( data, isCached ), $attributionMsg );
-	this.$popupContent.empty().append( $html );
+			.html( Tools.i18nHtml( scoreMsgKey, $scorePercent ) );
+
+	return $( '<div>' )
+		.append( addedMsg, getCommentContent( data ), $attributionMsg );
+}
 
+/**
+ * Show the revision popup based on the given token data, above the given element.
+ * Note that the English namespaces will normalize to the wiki's local namespaces.
+ * @param {Object} data As returned by Api.prototype.getTokenInfo().
+ * @param {jQuery} $target Element the popup should be attached to.
+ * @param {boolean} [isCached] Whether the comment promise is available, cached.
+ */
+RevisionPopupWidget.prototype.show = function ( data, $target, isCached ) {
```

Comment:
I'm thinking it might be better at this point to use the fact that the request is a promise for this.

That is, instead of having ifs and elses here, we can have the widget request the API for the promise, if it's uncached, set pending state, and unset it when the data is available. From that moment on, the promise is immediately-resolved, so doing it over and over would mean that we don't show loading animation but still request the cached promise.

Does this make sense? It just seemed a bit clearer usage (widget asks API for promise and shows) rather than something else calls the API and then calls the popup show function. This part used to work logically when we expected a single html "dump" (with or without comment) and repeatedly called `show` from outside the system -- but there's no real need for this now that we have a promise, and that we need to show the entire data of that promise in this widget. 

Does this make sense? I am thinking the widget can basically be more standalone by doing its own process, and whatever else calls it just toggles it on and off without having to know anythinga bout whether the promise exists or is cached, etc.

Reply:
I think so... and it sounds great! I was under some weird impression that the widgets shouldn't have knowledge of the API (like a view), and App was sort of the controller (gets data from API and gives it to the view) based on page interaction. If I'm following you right, I think all of [this logic](https://github.com/wikimedia/WhoWroteThat/blob/8f199808e92e616192cbbee1e3a4ddf9adb498e1/src/App.js#L198-L219) will need to be moved to RevisionPopupWidget as well? My only concern is this seems like a lot more refactoring. Maybe we could do it with a follow-up commit?

Reply:
Yeah API is a bit of a limbo state; usually, it is the Controller (in an MVC) that does it, then uses the API response to change the model -- and the widgets only listen to the model and invoke the controller.

So, we can send the widget the controller, and just ask for the API promise instead of having to refactor both the widget and the controller. Also, you're pointing out another logic weirdness where the App (somewhat the initialization-style-controller?) is calling the controller, which is a little meh, but might be okay if it's consistent about where we invoke what. It would probably be better for the controller itself to do that method, and then have `App` call that method.

So with all that said -- I do agree that that's quite a bit of a deeper change. I am good with doing this as a later refactor, but maybe we should document it with a ticket, just so we don't forget after the holidays? I think that beyond making things more readable and "proper" (arguably :D ) the refactor can prevent some potential sync issues that could happen.

Let's open another ticket, and I'll re-review this patch as-is without refactoring.

Reply:
(btw, unrelated to this change, but an addendum to my comment about the MVC -- part of the mismatch here is that while we have a Model, it's for the general system and not so much for the individual widget state. That might be something we could look at too, see if it makes sense to have the popup listen to the model for its data/state (am I in loading state, do I have a cached data, when the fetch happens, etc) rather than do that internal logic inside itself? It's not a must, but if we're already going to look at the way it works and simplify it, it might be a good point to revisit.)